Slotfilling interfaces, slotfilling implementation for Caila and CailaIntentActivator improvements

Denire opened the pull request:
This pull request will add slotfilling feature to JAICF. It also will create first slotfilling implementation for Caila NLU provider along will minor improvements in Caila NLU features.

A bit about improvements in Caila NLU:
1. Caila will now return a list of possible intents
2. Caila will now return more detailed NLU data - phrase markup and all the entities in phrase.

morfeusys commented on core/src/main/kotlin/com/justai/jaicf/slotfilling/SlotFiller.kt:
```diff
@@ -0,0 +1,33 @@
+
+/**
+ * This interface can be implemented to create a special way to handle filling some slots.
+ * It can be useful, for example, if you want to create channel-specific slot filling,
```
to create channel-specific slot filling **reactions**

morfeusys commented on core/src/main/kotlin/com/justai/jaicf/slotfilling/SlotFiller.kt:
```diff
@@ -0,0 +1,33 @@
+interface SlotFiller {
+
+    /**
+     * Tells if it's possible to fill custom this slot with your own code.
```
Please review this comment

morfeusys commented on core/src/main/kotlin/com/justai/jaicf/slotfilling/SlotFiller.kt:
```diff
@@ -0,0 +1,33 @@
+     * Tells if it's possible to fill custom this slot with your own code.
+     *
+     * @param name - name of slot to be filled
+     * @return true if can handle, no if can not
```
Please review this comment

morfeusys commented on core/src/main/kotlin/com/justai/jaicf/slotfilling/SlotFiller.kt:
```diff
@@ -0,0 +1,33 @@
+    /**
+     * Handles process of filling specific slot using current request's channel reactions
+     *
+     * @param name - name of slot to handle
```
name of slot to be handled

morfeusys commented on core/src/main/kotlin/com/justai/jaicf/slotfilling/SlotFillingResult.kt:
```diff
@@ -0,0 +1,51 @@
+ * @see [com.justai.jaicf.activator.Activator]
+ * */
+data class SlotFillingFinished(
+    val activatorContext: ActivatorContext
```
Please describe this parameter

morfeusys commented on activators/caila/src/main/kotlin/com/justai/jaicf/activator/caila/slotfilling/CailaSlotfillingHelper.kt:
```diff
@@ -0,0 +1,148 @@
+import com.justai.jaicf.slotfilling.*
+
+
+internal class CailaSlotfillingHelper(
```
Camel-case here please

morfeusys commented on activators/caila/src/main/kotlin/com/justai/jaicf/activator/caila/CailaIntentActivator.kt:
```diff
@@ -4,27 +4,49 @@ import com.justai.jaicf.activator.Activator
 
+    override val name = "cailaIntentActivator"
+
+    private val slotFillingHelper = CailaSlotfillingHelper(client)
```
How can I override a default slot filling settings?

morfeusys commented on activators/caila/src/main/kotlin/com/justai/jaicf/activator/caila/slotfilling/CailaSlotfillingHelper.kt:
```diff
@@ -0,0 +1,148 @@
+        return ctx
+    }
+
+    private fun getSlotfillingContext(botContext: BotContext): CailaSlotFillingContext? {
```
Why do you need a client ID again?

morfeusys replied:
Fix naming please

morfeusys commented on activators/caila/src/main/kotlin/com/justai/jaicf/activator/caila/slotfilling/CailaSlotfillingHelper.kt:
```diff
@@ -0,0 +1,148 @@
+                            reactions.sayRandom(*slot.prompts?.toTypedArray() ?: return SlotFillingSkipped)
+                        }
+                    }
+                    saveSlotFillingContext(botContext, ctx)
```
Did you ever test this persistence with some context managers like mongo or mapdb?

morfeusys commented on activators/caila/src/main/kotlin/com/justai/jaicf/activator/caila/slotfilling/CailaSlotfillingHelper.kt:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    companion object {
+        private fun getSlotfillingKey(clientId: String) = "$clientId-slotfilling-context"
```
Fix naming please

morfeusys commented on core/src/main/kotlin/com/justai/jaicf/slotfilling/SlotFillingResult.kt:
```diff
@@ -0,0 +1,54 @@
+ * This result indicates that slot filling session is finished and returns activator context,
+ * which will be passed to scenario.
+ *
+ * @param activatorContext initially stored activator context. New slots can be added to this context to be available in scenario.
```
"New slots SHOULD be added"

Denire commented:
@morfeusys Thank you for such useful comments!

After your last review I've changed or fixed:
1. Persistence with mapdb / mongo
2. Naming - use camelcase (SlotFilling instead of Slotfilling)
3. Added missing javadocs, or fixed bad javadocs
4. CailaIntentActivator intent selection logic (now by context + confidence)
5. Interface naming (SlotFiller -> SlotFillingReactionsProcessor), which now has methods canProcess and process

morfeusys commented on core/src/main/kotlin/com/justai/jaicf/slotfilling/SlotFillingBotContextExt.kt:
```diff
@@ -0,0 +1,15 @@
+
+internal fun BotContext.getSlotFillingActivator() = session[SLOTFILLING_ACTIVATOR_KEY] as? String
+
+internal fun BotContext.setSlotFillingIsFinished() = session.remove(SLOTFILLING_ACTIVATOR_KEY)
```
By fact, this function doesn't set anything. I'd rename it to `finishSlotFilling`

morfeusys commented on activators/caila/src/main/kotlin/com/justai/jaicf/activator/caila/slotfilling/CailaSlotFillingHelper.kt:
```diff
@@ -0,0 +1,152 @@
+                    return SlotFillingInterrupted
+                }
+                if (slot.name !in filled) {
+                    if (slotFillingReactionsProcessor != null && slotFillingReactionsProcessor.canProcess(slot.name)) {
```
`slotFillingReactionsProcessor?.canProcess(slot.name) == true`?

morfeusys commented on activators/caila/src/main/kotlin/com/justai/jaicf/activator/caila/slotfilling/CailaSlotFillingHelper.kt:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    companion object {
+        private const val SLOTFILLING_CONTEXT_KEY = "com/justai/jaicf/activator/caila/slotfilling/context"
```
Why not to use dots instead of slashes?

morfeusys commented on activators/caila/src/main/kotlin/com/justai/jaicf/activator/caila/slotfilling/CailaSlotFillingContext.kt:
```diff
@@ -0,0 +1,38 @@
+import com.justai.jaicf.context.ActivatorContext
+import kotlinx.serialization.Serializable
+
+@Serializable
```
Are you sure we need kotlinx serializable here?

Denire commented:
@morfeusys Thank you again, i'll sum up here:
1. Dots can be used in mongodb key only since 3.6, therefore I decided to use slashes to permit clients using mongo <3.6
2. SlotFillingReactionsProcessor was renamed to SlotReactor with methods canReact(...) and react(...), with improved javadocs.
3. Other points from your comments were fixed.

morfeusys reviewed (approved):
Good job @Denire !
